Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -331,6 +345,148 @@ public LibraryDependenciesResult(@Nullable final String[] libraries,
+      final Collection<String> rootsToAddToLib = new THashSet<String>();
+
+      for (final String path : libraries) {
+        if (path != null) {
+          folderPaths.add(PathUtil.getParentPath(FileUtil.toSystemIndependentName(path)));
+        }
+      }
+
+      outer:
+      for (final String path : folderPaths) {
+        final VirtualFile vFile = LocalFileSystem.getInstance().findFileByPath(path);
+        if (!path.startsWith(sdk.getHomePath() + "/") && (vFile == null || !fileIndex.isInContent(vFile))) {
+          for (String configuredPath : rootsToAddToLib) {
+            if (path.startsWith(configuredPath + "/")) {
+              continue outer; // folderPaths is sorted so subfolders go after parent folder
+            }
+          }
+          rootsToAddToLib.add(path);
+        }
+      }
+
+
+      final Set<Module> affectedModules = new THashSet<Module>();
+      final Module[] modules = ModuleManager.getInstance(project).getModules();
+      for (final Module module : modules) {
+        if (DartSdkGlobalLibUtil.isDartSdkGlobalLibAttached(module, sdk.getGlobalLibName())) {
+          for (final VirtualFile contentRoot : ModuleRootManager.getInstance(module).getContentRoots()) {
+            // if there is a pubspec, skip this contentRoot
+            if (contentRoot.findChild(PubspecYamlUtil.PUBSPEC_YAML) != null) continue;
+            affectedModules.add(module);
+          }
+        }
+      }
+
+      final Library library = createDartPackagesLibrary(project, rootsToAddToLib);
+
+      for (final Module module : affectedModules) {
+        final ModifiableRootModel modifiableModel = ModuleRootManager.getInstance(module).getModifiableModel();
+        try {
+          OrderEntry existingEntry = null;
+          for (final OrderEntry entry : modifiableModel.getOrderEntries()) {
+            if (entry instanceof LibraryOrderEntry &&
+                LibraryTablesRegistrar.PROJECT_LEVEL.equals(((LibraryOrderEntry)entry).getLibraryLevel()) &&
+                DartPackagesLibraryType.DART_PACKAGES_LIBRARY_NAME.equals(((LibraryOrderEntry)entry).getLibraryName())) {
+              existingEntry = entry;
+              break;
+            }
+          }
+
+          final boolean contains = existingEntry != null;
+          final boolean mustContain = affectedModules.contains(module);
+
+          if (contains != mustContain) {
+            if (mustContain) {
+              modifiableModel.addLibraryEntry(library);
+            }
+            else {
+              modifiableModel.removeOrderEntry(existingEntry);
+            }
+          }
+
+          if (modifiableModel.isChanged()) {
+            modifiableModel.commit();
+          }
+        }
+        finally {
+          if (!modifiableModel.isDisposed()) {
+            modifiableModel.dispose();
+          }
+        }
+      }
+    }
+  }
+
+  private static Library createDartPackagesLibrary(@NotNull final Project project,
```

Comment:
Adding a TODO to this effect. It is very similar, but different.  It will take a bit of rework to use DartLibInfo.  I don't want to potentionally break the code before I leave on a leave, and you leave on vacation.